akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphPartitionSpec.scala

```diff
@@ -159,6 +159,60 @@ class GraphPartitionSpec extends StreamSpec("""
       p1Sub.expectCancellation()
     }
 
+    "handle upstream completes and downstream cancel" in assertAllStagesStopped {
+      val c1 = TestSubscriber.probe[String]()
+      val c2 = TestSubscriber.probe[String]()
+
+      RunnableGraph
+        .fromGraph(GraphDSL.create() { implicit b =>
+          val partition = b.add(Partition[String](2, {
+            case s if s == "a" || s == "b" => 0
+            case _                         => 1
+          }))
+          Source(List("a", "b", "c", "d")) ~> partition.in
+          partition.out(0) ~> Sink.fromSubscriber(c1)
+          partition.out(1) ~> Sink.fromSubscriber(c2)
+          ClosedShape
+        })
+        .run()
+
+      c1.request(10)
+      c2.request(1)
+      c1.expectNext("a")
+      c1.expectNext("b")
+      c2.expectNext("c")
+      c2.cancel()
+      c1.expectComplete()
```

Comment:
Nice example

Reply:
Was this racy given that cancel and completion come from separate directions, or did it consistently repeat the problem?

Reply:
I don't think it was a race, throw the exception when I ran it (but I didn't repeat much)